[Chrus Chong Kai Ngee] iP

chrus-chong opened the pull request:
## Duke Productivity Chat-bot

> Always deliver more than expected  - [Larry Page](https://www.forbes.com/profile/larry-page/?sh=12c1273b7893)

The Duke Productivity Chat-bot frees your mind of having to remember events, deadlines, and tasks you need to do. It's :
* Text-based
* Easy to learn
* ~~FAST~~ SUPER FAST to use

## How it works
1. Download the file [here](https://github.com/chrus-chong/ip/releases/).
2. Navigate to the directory containing the file.
3. To run the file, type in `java -jar Duke.Productivity.Chat-Bot.jar` in the command line.
4. Learn to use Duke using the `help` command.

## Features
- [x] Managing deadlines with due date and time
- [x] Managing events with date as well as start and end time
- [x] Managing to dos with no specified date
- [ ] GUI Support


## Miscellaneous
If you are a Java Programmer, you can use it to practice Java too. 🚀 Here's the ```java main``` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

chrus-chong committed:
Complete basic level 1 functionalities (to add aesthetic features later on) and omit class files in .gitignore

chrus-chong committed:
Add and List functionality working and create Task class

chrus-chong committed:
Add status icon and done functionality

chrus-chong committed:
Add ToDo, Event and Deadline tasks and user guide command

chrus-chong committed:
Warning Fixes for Duke.java

chrus-chong committed:
Enable automated text UI testing

chrus-chong committed:
Create DukeException class to handle erroneous input

chrus-chong committed:
Add delete functionality and exception handling

chrus-chong committed:
Update tests to include delete tests

chrus-chong committed:
Refactoring of code from main method to run method

chrus-chong committed:
Further refactoring of code in run method

chrus-chong committed:
Add JavaDocs and minor bug fixes for deadlineCommand and eventCommand

chrus-chong committed:
Update EXPECTED.TXT

chrus-chong committed:
Change all methods in Duke to private in line with OOP Principles

chrus-chong committed:
Resolve all warnings for Levels 1-6

chrus-chong committed:
Bug fix for excess arguments to "delete" command

chrus-chong committed:
Add methods to read from, write to, and create file upon changes to tasklist

chrus-chong committed:
Remove unnecessary enum declaration

chrus-chong committed:
Bug fixes for readFromFile so that done tasks are reflected properly

chrus-chong committed:
Remove enum declaration and add warning fixes

chrus-chong committed:
Add functionality to store dates and time as LocalDate and LocalTime respectively

chrus-chong committed:
Merge branch 'branch-Level-7'

chrus-chong committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

chrus-chong committed:
Refactor code into the following classes: Ui, Storage, Parser, TaskList, Command

chrus-chong committed:
Divide classes into the following packages: duke, commands, exceptions, parser, storage, tasks, ui

chrus-chong committed:
Add JUnit tests for Parser and all Task classes

chrus-chong committed:
Correct typo in ListCommand

chrus-chong committed:
Remove META-INF folder

chrus-chong committed:
Update .gitignore

chrus-chong committed:
Add JavaDocs to all classes in duke folder

chrus-chong committed:
Add FindCommand to allow search by keywords

chrus-chong committed:
Format files to follow coding standard

Files not uniformly formatted.
Format files using CheckStyle to ensure uniformaity.

chrus-chong committed:
Merge branch 'branch-A-JavaDoc'

chrus-chong committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/commands/Command.java
#	src/main/java/duke/commands/DeadlineCommand.java
#	src/main/java/duke/commands/DeleteCommand.java
#	src/main/java/duke/commands/DoneCommand.java
#	src/main/java/duke/storage/Storage.java
#	src/main/java/duke/ui/Ui.java

chrus-chong committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/tasks/Deadline.java
#	src/main/java/duke/tasks/Event.java

chrus-chong committed:
Merge branch 'add-gradle-support'

chrus-chong committed:
Implement gradle to build, run, and run tests for Duke

No Javadocs for Duke, Ui, Command, FindCommand, ListCommand, and all Task class.
Cause gradle build to fail.
build.gradle file list the mainClassName as "seedu.duke.Duke".

Gradle build task unable to run.
Gradle run task unable to run.

Add Javadocs for all relevant files.
Change mainClassName from "seedu.duke.Duke" to "duke.Duke"

chrus-chong committed:
Merge branch 'branch-A-Gradle'

chrus-chong committed:
Create GUI Launcher that runs Hello World

Duke main method returns JavaFX runtime components are missing error.

chrus-chong committed:
Add GUI for Duke to echo user command

chrus-chong committed:
Add spacing between each DialogBox

chrus-chong committed:
Refactor Commands to return strings and add GUI functionality

Commands use System.out.print to log execution.
GUI implementation does not support input through System.out.print.

GUI implementation only accepts Strings as input.

Modify the execute method in all Command classes to return a string.

chrus-chong committed:
Bug fixes for deleteCommand and refining GUI

DeleteCommand writes to files before removing task.
No welcome message for user present.

Does not remove Task upon DeleteCommand.
Lack of welcome message fails to set context for user.

Fix DeleteCommand to remove task before write to files.
Welcome message added upon initialisation of MainWindow

{why it is done that way}

{any other relevant info}

chrus-chong committed:
Merge branch 'branch-Level-10'

Pulpy-y commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,66 @@
+            fxmlLoader.setRoot(this);
+            fxmlLoader.load();
+            Circle userCir = new Circle(47, 47, 43);
+            this.displayPicture.setClip(userCir);
```
Good to use circle as a clip to make the icon more intuitive 👍

Pulpy-y commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,28 @@
+/**
+ * Duke class used to run the duke.Duke chat-bot.
+ * Contains methods that
+ * (i)    initialises the Ui class and runs the chat-bot.
```
Maybe no need the (i) ?

Pulpy-y commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    /**
+     * You should have your own function to generate a response to user input.
```
Can tidy up the code and clean java doc upon finishing all codes.

Pulpy-y commented on src/main/java/duke/commands/ByeCommand.java:
```diff
@@ -0,0 +1,26 @@
+     * execute() method in ByeCommand to print leaving message.
+     *
+     * @param des   the user input into the Duke chat-box.
+     * @param tList the TaskList object used to keep track of all tasks.
```
I think it is better to capitalize the sentence after parameter name.

chrus-chong replied:
@Pulpy-y this was something that i overlooked. Thank you! will get to it soon

Pulpy-y commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,117 @@
+                            + "Please refer to the available commands using the \"help\" command");
+                } else {
+                    Command c = null;
+                    if (command.equals("bye")) {
```
I think can replace if-else statements with switch case?

chrus-chong committed:
Refine javadocs comments and update helpCommand

Javadoc header comments do not start with action verb.
helpCommand only contains functionalities up till delete.

Javadoc header comments need to be standardised in line with Git convention.
helpCommand needs to be updated to include new methods of formatting tasks input and new commands such as find and help.

Javadoc header comments rephrased to start with action verb.
Add in new commands and methods of formatting tasks in helpCommand .

Changing Javadoc comments to match Git convention allows for standardization.
helpCommand contains the most up-to-date information on all commands and formatting.

chrus-chong committed:
Refactor Ui, GraphicUi, DoneCommand and DeleteCommand for readability

Implementation of Ui and GraphicUi is unneccessarily long.
Each line of Javadoc comments in parameter section starts with non-capitalised letter.
Happy path nested in else statement for DoneCommand and DeleteCommand classes.

Ui and Graphic Ui contain excess code that can be shortened.
Parameter section of Javadoc comments does not conform to Git convention.
DoneCommand and DeleteCommand code is less intuitive and easy to understand.

Replace if-else statements in Ui and GraphicUi with switch-case statements.
Replace first letter of each line of Javadoc comments in parameter section with capitalised letter.
Refactor DoneCommand and DeleteCommand such that happy path is unindented as much as possible.

Ui and GraphicUi more readable.
Javadoc comments adhere to Git convention.
DoneCommand and DeleteCommand more readable.

chrus-chong committed:
Use assert to ensure commands do not take in illegal descriptions

Command classes may take in null descriptions or illegal characters such as '/'.

Create method to check for valid description in abstract Command class.

Ensure that command classes do not take in illegal descriptions.

chrus-chong committed:
Merge pull request #3 from chrus-chong/branch-A-Assertions

Use assert to ensure commands do not take in illegal descriptions

chrus-chong committed:
Merge pull request #4 from chrus-chong/branch-A-CodeQuality

Refactor Ui, GraphicUi, DoneCommand and DeleteCommand for readability

chrus-chong committed:
Add DoAfter task functionality

Tasks are standalone and do not reference one another.

Tasks in the real world are often linked. May need to complete one task
only after the completion of another task.

Add DoAfter task functionality to keep track of tasks that can only be
done after certain tasks.

Duke better able to handle various tasks.

chrus-chong committed:
Merge branch 'branch-DoAfterTask'

chrus-chong committed:
Bug fixes for Command class using assert feature

Gradle does not run with assertions enabled.
With assertions enabled, deadline and event command fails.

Duke does not function as it should without deadline and event
command.

Remove checks for '/' character in checkValidDes() method in
Command class.

Both event and deadline command functioning normally.

chrus-chong committed:
Refactor code to follow SLAP principles in Command and Parser classes

Code in Command and Parser classes do not follow SLAP.

Practicing single level of abstraction makes code neater.

Refactor Command and Parser classes such that relevant code is
abstracted.

Keeps in line with SLAP to ensure neatness and maintainability.

chrus-chong committed:
Update README and add app screenshot in /docs

ramapriyan912001 commented on src/main/java/duke/commands/Command.java:
```diff
@@ -0,0 +1,74 @@
+ * (ii)   print out the number of tasks in a given TaskList
+ * (iii)  counts the number of blank spaces in a string of text.
+ */
+public abstract class Command {
```
Prudent to have a command class from which commands can inherit! Great job!

ramapriyan912001 commented on src/main/java/duke/commands/Command.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    /**
+     * Returns Integer type object that represents the number of blank spaces in the user input.
```
Perhaps there can be a better way of achieving your goal than utilising a function like this which is a bit counterintuitive?

ramapriyan912001 commented on src/main/java/duke/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,60 @@
+     * @throws DukeException If input for Delete command is not properly formatted.
+     */
+    @Override
+    public String execute(String des, TaskList tList) throws DukeException {
```
Very well modularised. Good job!

ramapriyan912001 commented on src/main/java/duke/commands/DoneCommand.java:
```diff
@@ -0,0 +1,70 @@
+     * @throws DukeException If task has been completed, too many arguments are provided,
+     * or if input number is not valid.
+     */
+    private Task markTaskAsDone(String des, TaskList tList) throws DukeException {
```
Perhaps a better name would be "markAsDone" seeing as this is the task class itself.

ramapriyan912001 commented on src/main/java/duke/commands/DoneCommand.java:
```diff
@@ -0,0 +1,70 @@
+        ArrayList<Task> tasks = tList.getTaskList();
+        String stringNumber = des.substring(des.lastIndexOf(' ') + 1);
+        int num = Integer.parseInt(stringNumber);
+        if (num <= 0 || num > tasks.size()) {
```
Perhaps you could use a constant for these messages instead?

ramapriyan912001 commented on src/main/java/duke/commands/DoneCommand.java:
```diff
@@ -0,0 +1,70 @@
+            throw new DukeException("The input number is not a valid task number \n"
+                    + "Please refer to the task list using the \"list\" command");
+        }
+        if (this.countSpaces(des) > 1) {
```
Perhaps a comment here to explain the logic of the bloc?

ramapriyan912001 commented on src/main/java/duke/commands/HelpCommand.java:
```diff
@@ -0,0 +1,65 @@
+     * @return String representation of HelpCommand.
+     */
+    @Override
+    public String execute(String des, TaskList tList) {
```
Very helpful help guide for the user. Great work Chrus!

ramapriyan912001 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,59 @@
+     * @param des User input into the Duke chat-box.
+     * @return keyword present in user input
+     */
+    public String checkForKeyword(String des) {
```
The method could perhaps be SLAP-ed harder?

ramapriyan912001 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,222 @@
+     * @return ArrayList of LocalTime objects containing start and end time of an event.
+     * @throws DukeException When date is not properly formatted.
+     */
+    public static ArrayList<LocalTime> extractTimeEvent(String des) throws DukeException {
```
Good to modularise this.

ramapriyan912001 commented on src/main/java/duke/ui/GraphicUi.java:
```diff
@@ -0,0 +1,90 @@
+     * @param des User input into the Duke chat-box.
+     * @return String object representing the corresponding response from Duke.
+     */
+    public String run(String des) {
```
Very clear flow of execution, great job!

ramapriyan912001 reviewed (commented):
Great job Chrus! Just some minor points about magic literals.

chrus-chong committed:
Refactor code in line with feedback from CS2103 bot

chrus-chong committed:
Resolve minor checkstyle violation in Parser.java

chrus-chong committed:
Modify test cases to pass
